### PR TITLE
Hide dom elements before removal is finished. Resolves #103.

### LINKED ISSUE #103: App route link without reload or showing two routes simultaneously
**Describe the bug or question**
I want to add a link from my default route to a route with an app. However, using the `singleSpaNavigate` code in the [`navigateToUrl` docs](https://single-spa.js.org/docs/api/#navigatetourl) causes a reload. When I prevent the reload, it briefly shows both routes at the same time before navigating.

How can I prevent a reload without showing both routes simultaneously?

**To Reproduce**
1. Pull [this example repo](https://github.com/garethbjohnson/single-spa-buggy-demo)
2. Run `yarn`
3. Run `yarn start` in both the `root-config` and `app` folders in different terminals
4. Go to `localhost:9000`
5. Click the first link to see the reload
6. Click the second link to see both routes briefly appear simultaneously

These are the steps I took to make the repo:
1. Followed the [quick start docs](https://single-spa.js.org/docs/getting-started-overview#quick-start) to make a root config with Layout Engine and an app with React
2. Used this template in the root config `index.ejs`:
```html
<template id="single-spa-layout">
  <single-spa-router>
    <route default>
      <h1>Home</h1>
      <ul>
        <li>
          <a href="/app" onclick="singleSpaNavigate()">Link 1 (causes reload)</a>
        </li>
        <li>
          <a href="#" onclick="singleSpaNavigate('/app')">Link 2 (briefly shows two routes simultaneously)</a>
        </li>
      </ul>
    </route>

    <route path="app">
      <h1>App</h1>
      <application name="@buggy-demo/app"></application>
    </route>
  </single-spa-router>
</template>
```

**Expected behavior**
I expected clicking each link to show the app route without a reload or showing both routes at once.

**Screenshots and/or console output**
Here is a demo of the buggy behaviour, where Link 1 uses the documented `singleSpaNavigate` approach and Link 2 uses `href="#"` to prevent a reload:

[bug demo](https://user-images.githubusercontent.com/5997302/103928285-873bbe00-5113-11eb-8afd-4a7a5c313082.mov)

You can briefly see the app heading below the default route content.

### PR DESCRIPTION
**Pull request by joeldenning**
See #103. One thing to note is that none of the many snapshots had to change for this, which gave me a lot of confidence that it will be a safe change to release, despite being somewhat large.

We perform dom changes before unmounting finishes because the container elements for applications are needed before we let single-spa proceed to mounting applications. Previously, all dom elements were visible at the same time as the dom elements for the old route - this hides them and then shows them once it's ready.

### CODE REVIEW

File: src/browser/constructLayoutEngine.js [hunk 1]
```diff
@@ -170,29 +172,41 @@ export function constructLayoutEngine({
...
+          if (applicationElement && applicationElement.isConnected) {
+            applicationElement.parentNode.removeChild(applicationElement);
+          }
         }
-      }
-    );
+      );
+    }
+
+    pendingMakeVisibles.forEach((makeVisible) => makeVisible());
```
Comment (joeldenning) [inline src/browser/constructLayoutEngine.js:208]: This approach has the downside of not making any of the dom elements visible until the slowest application has finished mounting. I think it's okay, though

Comment (filoxo) [reply src/browser/constructLayoutEngine.js:208]: Could that mess with animations for transitioning in/out?

Comment (joeldenning) [reply src/browser/constructLayoutEngine.js:208]: Yes transitions in would be affected. Not transitions out, though.

**Comment by joeldenning** [comment]
I thought over this again in relationship to the following code: https://github.com/single-spa/single-spa/blob/b35d73ad9de433f529f4e51767c4e8b46d8f892b/src/navigation/reroute.js#L134

Taking that into account, I think this is the wrong implementation approach. Closing this PR - I'll put together a different one with a better implementation approach.

**Pull request closed without merging**